src/box/errcode.h

```diff
@@ -378,6 +424,18 @@ tnt_errcode_desc(uint32_t errcode)
 	return box_error_codes[errcode].errdesc;
 }
 
+/**
+ * Return the error fields info for the errorcode.
+ * Result is NULL if errorcode is unknown.
+ */
+static inline const struct errcode_record *
+tnt_errcode_record(uint32_t errcode)
+{
+	if (!tnt_errcode_is_defined(errcode))
+		return NULL;
```

Comment:
I think `tnt_errocode_record` should return a special record for the unknown error. Also, I think that `tnt_errcode_desc` and `tnt_errcode_str` should call this function.

Note if you decide to define the special record in the function body, you should make the function `extern` to avoid multiple definitions.

Reply:
Done. Now related hunks looks likes this:


```diff
  */

 ENUM(box_error_code, ERROR_CODES);
-extern struct errcode_record box_error_codes[];
+extern const struct errcode_record box_error_codes[];
+
+/** Record for unknown errocode. */
+extern const struct errcode_record errcode_record_unknown;

-static inline bool
-tnt_errcode_is_defined(uint32_t errcode)
+/** Return the error fields info for the errorcode. */
+static inline const struct errcode_record *
+tnt_errcode_record(uint32_t errcode)
 {
-       return errcode < box_error_code_MAX &&
-              box_error_codes[errcode].errstr != NULL;
+       if (errcode >= box_error_code_MAX ||
+           box_error_codes[errcode].errstr == NULL)
+               return &errcode_record_unknown;
+       return &box_error_codes[errcode];
 }

 /** Return a string representation of error name, e.g. "ER_OK". */
 static inline const char *
 tnt_errcode_str(uint32_t errcode)
 {
-       if (!tnt_errcode_is_defined(errcode)) {
-               /* Unknown error code - can be triggered using box.error() */
-               return "ER_UNKNOWN";
-       }
-       return box_error_codes[errcode].errstr;
+       return tnt_errcode_record(errcode)->errstr;
 }

 /** Return a description of the error. */
 static inline const char *
 tnt_errcode_desc(uint32_t errcode)
 {
-       if (!tnt_errcode_is_defined(errcode))
-               return "Unknown error";
-
-       return box_error_codes[errcode].errdesc;
+       return tnt_errcode_record(errcode)->errdesc;
 }

 #ifdef __cplusplus
diff --git a/src/box/errcode.c b/src/box/errcode.c
index 44f8860ba..9450b7205 100644
--- a/src/box/errcode.c
+++ b/src/box/errcode.c
        ERROR_CODES(ERRCODE_RECORD_MEMBER)
 };

+const struct errcode_record errcode_record_unknown = {
+       .errstr = "ER_UNKNOWN",
+       .errdesc = "Unknown error",
+};
@@ -611,31 +612,171 @@ test_undefined_error_code(void)
 {
        header();
 #ifdef TEST_BUILD
-       plan(4);
+       plan(12);
 #else
-       plan(2)
+       plan(6)
 #endif

+       const struct errcode_record *record;
        ok(strcmp(tnt_errcode_str(box_error_code_MAX), "ER_UNKNOWN") == 0);
        ok(strcmp(tnt_errcode_desc(box_error_code_MAX), "Unknown error") == 0);
+       record = tnt_errcode_record(box_error_code_MAX);
+       ok(strcmp(record->errstr, "ER_UNKNOWN") == 0);
+       ok(strcmp(record->errdesc, "Unknown error") == 0);
+       ok(record->errfields == NULL);
+       ok(record->errfields_count == 0);
 #ifdef TEST_BUILD
        ok(strcmp(tnt_errcode_str(ER_TEST_FIRST - 1), "ER_UNKNOWN") == 0);
        ok(strcmp(tnt_errcode_desc(ER_TEST_FIRST - 1), "Unknown error") == 0);
+       record = tnt_errcode_record(ER_TEST_FIRST - 1);
+       ok(strcmp(record->errstr, "ER_UNKNOWN") == 0);
+       ok(strcmp(record->errdesc, "Unknown error") == 0);
+       ok(record->errfields == NULL);
+       ok(record->errfields_count == 0);
 #endif

        check_plan();
        footer();
 }
```